src/node_url.cc

```diff
@@ -426,6 +428,106 @@ std::string FromFilePath(const std::string_view file_path) {
   return ada::href_from_file(escaped_file_path);
 }
 
+std::optional<std::string> FileURLToPath(Environment* env,
+                                         const ada::url_aggregator& file_url) {
+  if (file_url.type != ada::scheme::FILE) {
+    THROW_ERR_INVALID_URL_SCHEME(env->isolate());
+    return std::nullopt;
+  }
+
+  std::string_view pathname = file_url.get_pathname();
+#ifdef _WIN32
+  size_t first_percent = std::string::npos;
+  size_t pathname_size = pathname.size();
+  std::string pathname_escaped_slash;
+
+  for (size_t i = 0; i < pathname_size; i++) {
+    if (pathname[i] == '/') {
+      pathname_escaped_slash += '\\';
+    } else {
+      pathname_escaped_slash += pathname[i];
+    }
+
+    if (pathname[i] != '%') continue;
+
+    if (first_percent == std::string::npos) {
+      first_percent = i;
+    }
+
+    // just safe-guard against access the pathname
+    // outside the bounds
+    if ((i + 2) >= pathname_size) continue;
+
+    char third = pathname[i + 2] | 0x20;
+
+    bool is_slash = pathname[i + 1] == '2' && third == 102;
+    bool is_forward_slash = pathname[i + 1] == '5' && third == 99;
+
+    if (!is_slash && !is_forward_slash) continue;
+
+    THROW_ERR_INVALID_FILE_URL_PATH(
+        env->isolate(),
+        "File URL path must not include encoded \\ or / characters");
+
+    return std::nullopt;
+  }
+
+  std::string_view hostname = file_url.get_hostname();
+  std::string decoded_pathname =
+      ada::unicode::percent_decode(pathname_escaped_slash, first_percent);
+
+  if (hostname.size() > 0) {
+    // If hostname is set, then we have a UNC path
+    // Pass the hostname through domainToUnicode just in case
+    // it is an IDN using punycode encoding. We do not need to worry
+    // about percent encoding because the URL parser will have
+    // already taken care of that for us. Note that this only
+    // causes IDNs with an appropriate `xn--` prefix to be decoded.
```

Comment:
If you assume that xn-- is uncommon (or just very common), then a branch here on a search for xn-- might be a possible optimization.

(I don't particularly recommend it though.)

Reply:
I don't think it's *that* common. 👍